controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/PathGroup.java

```diff
@@ -0,0 +1,116 @@
+           "/application/v4/tenant/",
+           "/application/v4/tenant-pipeline/",
+           "/application/v4/tenant/{tenant}",
+           "/application/v4/tenant/tenant1/application/",
+           "/application/v4/tenant/{tenant}/application/{application}",
+           "/application/v4/tenant/{tenant}/application/{application}/deploying/{*}",
+           "/application/v4/tenant/{tenant}/application/{application}/instance/{instance}/job/{job}/{*}",
+           "/application/v4/tenant/{tenant}/application/{application}/environment/dev/{*}",
+           "/application/v4/tenant/{tenant}/application/{application}/environment/perf/{*}",
+           "/application/v4/tenant/{tenant}/application/{application}/environment/{environment}/region/{region}/instance/{instance}/global-rotation/override"),
+
+    /**
+     * Paths used for deployments by build service(s). Note that context is ignored in these paths as build service
+     * roles are not granted in specific contexts.
+     */
+    buildService("/zone/v1/{*}",
+                 "/application/v4/tenant/{tenant}/application/{application}/jobreport",
+                 "/application/v4/tenant/{tenant}/application/{application}/submit",
+                 "/application/v4/tenant/{tenant}/application/{application}/promote",
+                 "/application/v4/tenant/{tenant}/application/{application}/environment/prod/{*}",
+                 "/application/v4/tenant/{tenant}/application/{application}/environment/test/{*}",
+                 "/application/v4/tenant/{tenant}/application/{application}/environment/staging/{*}"),
+
+    /** Paths providing information about deployment status */
+    deployment("/badge/v1/{*}",
+               "/deployment/v1/{*}");
+
+    private final Set<String> pathSpecs;
+
+    PathGroup(String... pathSpecs) {
+        this.pathSpecs = Set.of(pathSpecs);
+    }
+
+    /** Returns path if it matches any spec in this group */
+    private Optional<Path> get(String path) {
+        return Optional.of(path).map(Path::new).filter(p -> pathSpecs.stream().anyMatch(p::matches));
+    }
+
+    /** All known path groups */
+    public static Set<PathGroup> all() {
+        return EnumSet.allOf(PathGroup.class);
+    }
+
+    /** Returns whether this group matches path in given context */
+    public boolean matches(String path, Context context) {
+        return get(path).filter(p -> {
+            boolean match = true;
+            String tenant = p.get("tenant");
+            if (tenant != null && context.tenant().isPresent()) {
+                match = context.tenant().get().value().equals(tenant);
+            }
+            String application = p.get("application");
+            if (application != null && context.application().isPresent()) {
+                match &= context.application().get().value().equals(application);
+            }
+            return match;
+        }).isPresent();
+    }
+
+    static {
+        // Ensure that all path spec sets are disjoint
```

Comment:
Shouldn't this just be a test?

Reply:
Sure, I just prefer to keep contracts within the class (similar requireSomething methods) when possible. If validation is a lot of code and/or needs other classes to do the validation, it can be moved to a test.